{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Author: Leonie KÃ¼chenhoff\n",
    "### Date: November 2022\n",
    "### Purpose of script: Write one list per tissue of tissue specific variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pacage import\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "from config import basedir, tissue_specdir, annotation_dir\n",
    "from itertools import product"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This base directory will be used:\n",
      " /g/steinmetz/project/leonie_crispr/03_data/02_rnaseq/snakemake/\n"
     ]
    }
   ],
   "source": [
    "#directory settings\n",
    "outdir = tissue_specdir\n",
    "annodir = annotation_dir\n",
    "print('This base directory will be used:\\n', basedir) \n",
    "os.chdir(basedir)\n",
    "# sample names\n",
    "names = ['028_pbs_R', '029_pbs_R', '030_nrch_R', '032_pbs_R', '033_nrch_R', '036_nrch_R',\n",
    "       '011_pbs', '012_nrch', '013_nrch', '014_nrch', '279_spry', '321_pbs', '333_pbs', '450_spry', '283_spry']\n",
    "varcallers = ['hc', 'pl', 'st']\n",
    "combinations = list(product(names, varcallers))\n",
    "\n",
    "paths = [f'filtered_tables/HL{i}.specific.annofilter.txt' for i in names]\n",
    "col1 = 'ad_h'\n",
    "col2 = 'ad_l' \n",
    "\n",
    "use_cols = ['Func.refGene', 'Gene.refGene',\n",
    "       'GeneDetail.refGene', 'ExonicFunc.refGene', 'AAChange.refGene',\n",
    "       'Otherinfo4', 'Otherinfo5', 'Otherinfo7', 'Otherinfo8']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "paths_anno = [f'{annodir}/HL_{i[1]}_{i[0]}.mm10_multianno.txt' for i in combinations]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "li = []\n",
    "for path in paths_anno:\n",
    "    anno = pd.read_csv(path, delimiter = '\\t', usecols = use_cols).rename(columns={'Otherinfo4':'chr', 'Otherinfo5':'pos', 'Otherinfo7':'ref', 'Otherinfo8':'alt'})\n",
    "    li.append(anno)\n",
    "\n",
    "anno = pd.concat(li, axis=0, ignore_index=True)\n",
    "anno = anno.drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sets(df):\n",
    "    setlist = np.arange(0, len(df))\n",
    "    heart = setlist[df['heart']]\n",
    "    liver = setlist[df['liver']]\n",
    "    return heart, liver"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_af(df, col1, col2):\n",
    "    # determine allele ferquency and total number of reads per variant and sample\n",
    "    mouse_ad = df[[col1, col2]]\n",
    "    allel1 = mouse_ad.applymap(lambda x: int(x.split(',')[0])).to_numpy()\n",
    "    allel2 = mouse_ad.applymap(lambda x: int(x.split(',')[1])).to_numpy()\n",
    "    af = allel2 / (allel1 + allel2)\n",
    "    af = np.nan_to_num(af, 0)\n",
    "    tot_reads = allel1 + allel2\n",
    "    return af, tot_reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "def annotate_duplicated(df_main, df_compare1, df_compare2):\n",
    "    '''\n",
    "    function to indicate wether variant occurs only in on sample or also other sample\n",
    "    '''\n",
    "    df_main_anno = pd.merge(df_main,df_compare1[['chr', 'pos', 'ref','alt', 'present_in_sample']], on =['chr', 'pos', 'ref','alt'], how = 'left')\n",
    "    df_main_anno['present_in_sample_y'] = df_main_anno['present_in_sample_y'].fillna('')\n",
    "    df_main_anno['present_in_sample_y'] = df_main_anno['present_in_sample_x'] +', ' + df_main_anno['present_in_sample_y']\n",
    "\n",
    "    df_main_anno = pd.merge(df_main_anno,df_compare2[['chr', 'pos', 'ref','alt', 'present_in_sample']], on =['chr', 'pos', 'ref','alt'], how = 'left')\n",
    "    df_main_anno['present_in_sample'] = df_main_anno['present_in_sample'].fillna('')\n",
    "    df_main_anno['present_in_sample'] = df_main_anno['present_in_sample_y'] +', ' + df_main_anno['present_in_sample']\n",
    "\n",
    "    df_main_anno = df_main_anno.drop(['present_in_sample_x', 'present_in_sample_y'], axis = 1)\n",
    "    return(df_main_anno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "use_cols = ['Func.refGene', 'Gene.refGene',\n",
    "       'GeneDetail.refGene', 'ExonicFunc.refGene', 'AAChange.refGene',\n",
    "       'Otherinfo4', 'Otherinfo5', 'Otherinfo7', 'Otherinfo8']\n",
    "\n",
    "sets = [['028_pbs_R', '029_pbs_R', '032_pbs_R'], ['033_nrch_R', '030_nrch_R', '036_nrch_R'],['014_nrch', '012_nrch', '013_nrch'], ['279_spry','450_spry', '283_spry'], ['011_pbs','333_pbs','321_pbs']]\n",
    "for set in sets:\n",
    "    '''\n",
    "    For each sample, read tissue specific files, annotate in which other samples these variants occur,\n",
    "    annotate and save as txt file\n",
    "    '''\n",
    "    names = set\n",
    "    print(names)\n",
    "    # read in tissue specific variants\n",
    "    file_dict = {}\n",
    "    for i in names:\n",
    "        path = f'filtered_tables/HL{i}.specific.annofilter.txt'\n",
    "        print(i)\n",
    "        print(path)\n",
    "        df = pd.read_csv(path, delimiter = '\\t')\n",
    "        file_dict[i] = df\n",
    "    venn_dict = {}\n",
    "    for i in names:\n",
    "        venn_dict[i] = get_sets(file_dict[i])   \n",
    "\n",
    "    annotated_heart_dict = {}\n",
    "    annotated_liver_dict = {}\n",
    "\n",
    "    for count, i in enumerate(names):\n",
    "        #heart\n",
    "        heart_spec = file_dict[i][file_dict[i]['heart'] == True]\n",
    "        af, tot_reads = get_af(heart_spec, col1, col2) \n",
    "        heart_spec['AF_h'], heart_spec['AF_l'] = af[:, 0], af[:, 1]\n",
    "        heart_spec['reads_h'], heart_spec['reads_l'] = tot_reads[:, 0], tot_reads[:, 1]\n",
    "        heart_spec_3 = heart_spec.iloc[np.where(af[:,1]==0)]\n",
    "        heart_file = heart_spec_3[['chr', 'pos', 'ref', 'alt', col1, col2, 'AF_h', 'AF_l', 'reads_h', 'reads_l', 'normed_h', 'normed_l']]\n",
    "\n",
    "        #liver\n",
    "        liver_spec = file_dict[i][file_dict[i]['liver'] == True]\n",
    "        af, tot_reads = get_af(liver_spec, col1, col2) \n",
    "        liver_spec['AF_h'], liver_spec['AF_l'] = af[:, 0], af[:, 1]\n",
    "        liver_spec['reads_h'], liver_spec['reads_l'] = tot_reads[:, 0], tot_reads[:, 1]\n",
    "        liver_spec_3 = liver_spec.iloc[np.where(af[:,0]==0)]\n",
    "        liver_file = liver_spec_3[['chr', 'pos', 'ref', 'alt', col1, col2, 'AF_h', 'AF_l', 'reads_h', 'reads_l', 'normed_h', 'normed_l']]\n",
    "        \n",
    "        # merge annotation with heart files\n",
    "        merged_h = pd.merge(heart_file, anno, how = 'left',on = ['chr', 'pos', 'ref', 'alt'])\n",
    "        merged_h.loc[:,'total_reads'] = merged_h.loc[:,'reads_h'] +  merged_h.loc[:,'reads_l']\n",
    "        merged_h.loc[:,'present_in_sample'] =  names[count]\n",
    "        annotated_heart_dict[names[count]] = merged_h\n",
    "\n",
    "        # merge annotation with liver files\n",
    "        merged = pd.merge(liver_file, anno, how = 'left',on = ['chr', 'pos', 'ref', 'alt'])\n",
    "        merged.loc[:,'total_reads'] = merged.loc[:,'reads_h'] +  merged.loc[:,'reads_l']\n",
    "        merged.loc[:,'present_in_sample'] =  names[count]\n",
    "        annotated_liver_dict[names[count]] = merged\n",
    "\n",
    "    # save heart files\n",
    "    heart1 = annotate_duplicated(annotated_heart_dict[names[0]], annotated_heart_dict[names[1]], annotated_heart_dict[names[2]])\n",
    "    heart1.to_csv(f'{outdir}/HC_PL_ST_heart_{names[0]}.txt', sep = '\\t', index = False)\n",
    "    heart2 = annotate_duplicated(annotated_heart_dict[names[1]], annotated_heart_dict[names[0]], annotated_heart_dict[names[2]])\n",
    "    heart2.to_csv(f'{outdir}/HC_PL_ST_heart_{names[1]}.txt', sep = '\\t', index = False)\n",
    "    heart3 = annotate_duplicated(annotated_heart_dict[names[2]], annotated_heart_dict[names[0]], annotated_heart_dict[names[1]])\n",
    "    heart3.to_csv(f'{outdir}/HC_PL_ST_heart_{names[2]}.txt', sep = '\\t', index = False)\n",
    "    \n",
    "    # save liver files\n",
    "    liver1 = annotate_duplicated(annotated_liver_dict[names[0]], annotated_liver_dict[names[1]], annotated_liver_dict[names[2]])\n",
    "    liver1.to_csv(f'{outdir}/HC_PL_ST_liver_{names[0]}.txt', sep = '\\t', index = False)\n",
    "    liver2 = annotate_duplicated(annotated_liver_dict[names[1]], annotated_liver_dict[names[0]], annotated_liver_dict[names[2]])\n",
    "    liver2.to_csv(f'{outdir}/HC_PL_ST_liver_{names[1]}.txt', sep = '\\t', index = False)\n",
    "    liver3 = annotate_duplicated(annotated_liver_dict[names[2]], annotated_liver_dict[names[0]], annotated_liver_dict[names[1]])\n",
    "    liver3.to_csv(f'{outdir}/HC_PL_ST_liver_{names[2]}.txt', sep = '\\t', index = False)\n"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e6f34ad77cfe09e92b789e4b2c65cd4206cf10540b9993c0effd6d86fa137da0"
  },
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
