{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Author: Leonie KÃ¼chenhoff\n",
    "### Date: October 2022\n",
    "### Purpose of script: Script to check list of variats in more detail.\n",
    "###             (Are there any gRNA Sequence similarities in proximity?)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.ticker import MaxNLocator\n",
    "from config import outdir, tissue_specdir, reference_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This base directory will be used:\n",
      " /g/steinmetz/project/leonie_crispr/03_data/02_rnaseq/snakemake/tissue_spec_vars\n"
     ]
    }
   ],
   "source": [
    "# sample names\n",
    "names = ['028_pbs_R', '029_pbs_R', '032_pbs_R','033_nrch_R', '030_nrch_R', '036_nrch_R', \n",
    "            '011_pbs', '012_nrch', '013_nrch', '014_nrch', '279_spry', '321_pbs', '333_pbs', '450_spry', '283_spry']\n",
    "# treatment in same order as sample names\n",
    "base_editor = ['pbs', 'pbs', 'pbs', '8e-NRCH', '8e-NRCH', '8e-NRCH',\n",
    "            'pbs', '8e-NRCH', '8e-NRCH','8e-NRCH', 'max-SpRY', 'pbs', 'pbs', 'max-SpRY', 'max-SpRY']\n",
    "mutation = ['r636q' if i[-1] == 'R' else 'p635l' for i in names]\n",
    "condition = [i+'_'+j for i,j in zip(base_editor, mutation)]\n",
    "#base directory\n",
    "basedir = tissue_specdir\n",
    "outdir = str(outdir + '/variant_surrounding/')\n",
    "print('This base directory will be used:\\n', basedir)    \n",
    "os.chdir(basedir)\n",
    "\n",
    "#path to refernce sequence\n",
    "ref_seq = reference_seq\n",
    "\n",
    "#gRNA + PAM sequence \n",
    "grna_spry = 'GAACGTAGCCGCTCTGGACCANRN'\n",
    "grna_nrch = 'GAACGTAGCCGCTCTGGACCANRCH'\n",
    "grnaR_nrch = 'GCACAGTCTCGAAGTCCAATGNRCH'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_reference_data(path_ref):\n",
    "    '''\n",
    "    function to read in reference sequence\n",
    "    '''\n",
    "    reference = SeqIO.to_dict(SeqIO.parse(path_ref, 'fasta'))\n",
    "    return reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in reference sequence\n",
    "reference  = read_reference_data(ref_seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in list of tissue specific variants to find location of variants\n",
    "heart_spec_dict = {}\n",
    "for i in names:\n",
    "    heart_spec_dict[i] = pd.read_csv(f'HC_PL_ST_heart_{i}.txt', delimiter='\\t')\n",
    "liver_spec_dict = {}\n",
    "for i in names:\n",
    "    liver_spec_dict[i] = pd.read_csv(f'HC_PL_ST_liver_{i}.txt', delimiter='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "chromosomes = ['chr1',\n",
    "                'chr2',\n",
    "                'chr3',\n",
    "                'chr4',\n",
    "                'chr5',\n",
    "                'chr6',\n",
    "                'chr7',\n",
    "                'chr8',\n",
    "                'chr9',\n",
    "                'chr10',\n",
    "                'chr11',\n",
    "                'chr12',\n",
    "                'chr13',\n",
    "                'chr14',\n",
    "                'chr15',\n",
    "                'chr16',\n",
    "                'chr17',\n",
    "                'chr18',\n",
    "                'chr19',\n",
    "                'chrX',\n",
    "                'chrY']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_refseq(position_dict, tissue, names):\n",
    "    '''\n",
    "    function to interate through positions of varaints and save reference sequence +/- 30 bases in table\n",
    "    '''\n",
    "    seq_dict = {}\n",
    "    for count, name in enumerate(names):\n",
    "        if name[-1] == 'R':\n",
    "            gRNA = grnaR_nrch\n",
    "        else:\n",
    "            if condition[count][0] == '8':\n",
    "                gRNA = grna_nrch\n",
    "            else:\n",
    "                gRNA = grna_spry\n",
    "        df_in = position_dict[name]\n",
    "\n",
    "        seq_list = []\n",
    "        for chr in chromosomes:\n",
    "            seq = reference[chr].upper()\n",
    "            positions = df_in[df_in['chr'] == chr]['pos'].to_list()\n",
    "            for i in positions:\n",
    "                seq_pos = seq[i-30: i + 30]\n",
    "                seq_list.append(str(seq_pos.seq))\n",
    "        seq_dict[name] = seq_list\n",
    "        df = pd.DataFrame({'gRNA':gRNA, 'genome_dna':seq_dict[name], 'chr':df_in['chr'], 'pos':df_in['pos'], 'ref':df_in['ref'], 'alt':df_in['alt'], 'normed_h':df_in['normed_h'], 'normed_l':df_in['normed_l']})\n",
    "\n",
    "        if condition[count][0] == \"p\":\n",
    "            df2 = pd.DataFrame({'gRNA':grna_nrch, 'genome_dna':seq_dict[name], 'chr':df_in['chr'], 'pos':df_in['pos'], 'ref':df_in['ref'], 'alt':df_in['alt'], 'normed_h':df_in['normed_h'], 'normed_l':df_in['normed_l']})\n",
    "            df = pd.concat([df2,df])\n",
    "\n",
    "        df.to_csv(f'{tissue}_genomic_surrounding_sample_{name}.txt', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "save_refseq(heart_spec_dict, 'heart', names)\n",
    "save_refseq(liver_spec_dict, 'liver', names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n",
      "sh: -c: line 4: syntax error: unexpected end of file\n"
     ]
    }
   ],
   "source": [
    "# costum perl script to align PAM sequence & gRNA to region around variant and output number of mismatches to best match in region\n",
    "#for i in names:\n",
    "#    os.system(f'''\n",
    "#    perl analysis/semi_global_alignment.pl -in heart_genomic_surrounding_sample_{i}.txt -o heart_genomic_surrounding_sample_{i}_analysis.txt &&\n",
    "#    perl analysis/semi_global_alignment.pl -in liver_genomic_surrounding_sample_{i}.txt -o liver_genomic_surrounding_sample_{i}_analysis.txt &&\n",
    "#    ''')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# read in results from perl script\n",
    "\n",
    "analysed_heart = pd.DataFrame()\n",
    "for count,i in enumerate(names):\n",
    "    analysed_sample = pd.read_csv(f'heart_genomic_surrounding_sample_{i}_analysis.txt',\n",
    "                                        delimiter = '\\t')\n",
    "    analysed_sample['sample'] = i\n",
    "    analysed_sample['tissue'] ='heart'\n",
    "    analysed_sample['condition'] = condition[count]\n",
    "    analysed_heart = pd.concat([analysed_heart, analysed_sample], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# exclude target-area\n",
    "analysed_heart = analysed_heart[~(analysed_heart['pos']>=53843223)& (analysed_heart['pos']<=53843268) & (analysed_heart['chr']>='chr19')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch_plot = analysed_heart[['mismtaches', 'condition', 'tissue']]\n",
    "mismatch_plot.to_csv(f'{outdir}/mismatch_plot.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors = ['bisque',\n",
    "          'orange',\n",
    "          'azure',\n",
    "          'steelblue',\n",
    "          'navy'\n",
    "          ]\n",
    "costum_palette_box = sns.color_palette(colors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mismatch_plot_r = mismatch_plot[mismatch_plot['condition'].isin(['pbs_r636q','8e-NRCH_r636q'])]\n",
    "mismatch_plot_p = mismatch_plot[mismatch_plot['condition'].isin(['pbs_p635l', '8e-NRCH_p635l','max-SpRY_p635l'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1,2, figsize = (12,5), sharey = True, sharex = True)\n",
    "plt.style.use('seaborn-colorblind')\n",
    "plt.style.use('seaborn-whitegrid')\n",
    "\n",
    "sns.histplot(data = mismatch_plot_r, x = 'mismtaches', hue = 'condition',multiple = 'dodge', palette = costum_palette_box[0:2],\n",
    "                stat = 'probability', ax = ax[0], binwidth = 1, binrange = (0.5,20.5), common_norm =False, \n",
    "                edgecolor = 'black', linewidth = 1.2, shrink = 0.7, alpha=1)\n",
    "sns.histplot(data = mismatch_plot_p, x = 'mismtaches', hue = 'condition',multiple = 'dodge',palette = costum_palette_box[2:6],\n",
    "                stat = 'probability', ax = ax[1], binwidth = 1, binrange = (0.5,20.5), common_norm =False, \n",
    "                edgecolor = 'black', linewidth = 1.2, shrink = 0.8, alpha=1)\n",
    "\n",
    "ax[0].get_legend().set_title('')\n",
    "ax[1].get_legend().set_title('')\n",
    "\n",
    "# replace labels\n",
    "new_labels = ['PBS', '8e-NRCH']\n",
    "for t, l in zip(ax[0].get_legend().texts, new_labels):\n",
    "    t.set_text(l)\n",
    "new_labels = ['PBS', '8e-NRCH', 'max-SpRY']\n",
    "for t, l in zip(ax[1].get_legend().texts, new_labels):\n",
    "    t.set_text(l)\n",
    "\n",
    "ax[0].set_xlabel('mismatches')\n",
    "ax[1].set_xlabel('mismatches')\n",
    "ax[0].set_xlim(-0.50,20)\n",
    "ax[0].set_ylim(-0.005,0.4)\n",
    "ax[0].set_ylabel('relative amount')\n",
    "ax[0].set_title('R636Q')\n",
    "ax[1].set_title('P635L')\n",
    "fig.tight_layout()\n",
    "plt.show()\n",
    "\n",
    "fig.savefig(f'{outdir}/tissue_spec_mismatches_separate.pdf', bbox_inches = 'tight')\n",
    "fig.savefig(f'{outdir}/tissue_spec_mismatches_separate.svg', bbox_inches = 'tight')"
   ]
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "e6f34ad77cfe09e92b789e4b2c65cd4206cf10540b9993c0effd6d86fa137da0"
  },
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
